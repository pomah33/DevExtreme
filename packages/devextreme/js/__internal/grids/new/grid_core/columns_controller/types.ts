import type { Format } from '@js/common';
import type { ColumnBase } from '@js/common/grids';

import type { DataObject, Key } from '../data_controller/types';

type InheritedColumnProps =
  | 'alignment'
  | 'dataType'
  | 'visible'
  | 'visibleIndex'
  | 'allowReordering'
  | 'trueText'
  | 'falseText'
  | 'caption';

export type Column = Pick<Required<ColumnBase>, InheritedColumnProps> & {
  dataField?: string;

  name: string;

  calculateCellValue: (this: Column, data: DataObject) => unknown;

  calculateDisplayValue: (this: Column, data: DataObject) => unknown;

  format?: Format;

  customizeText?: (this: Column, info: {
    value: unknown;
    valueText: string;
  }) => string;

  editorTemplate?: unknown;

  fieldTemplate?: unknown;
};

export type VisibleColumn = Column & { visible: true };

export interface Cell {
  value: unknown;

  displayValue: unknown;

  text: string;

  column: Column;
}

export interface DataRow {
  cells: Cell[];

  key: Key;

  data: DataObject;
}
