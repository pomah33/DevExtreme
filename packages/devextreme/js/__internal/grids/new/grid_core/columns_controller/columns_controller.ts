/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable spellcheck/spell-checker */
import formatHelper from '@js/format_helper';
import type { Subscribable, SubsGets, SubsGetsUpd } from '@ts/core/reactive/index';
import {
  computed, iif, interruptableComputed,
} from '@ts/core/reactive/index';

import { DataController } from '../data_controller/index';
import type { DataObject } from '../data_controller/types';
import { OptionsController } from '../options_controller/options_controller';
import type { ColumnProperties, ColumnSettings, PreNormalizedColumn } from './options';
import type { Column, DataRow, VisibleColumn } from './types';
import {
  getColumnIndexByName, normalizeColumns, normalizeVisibleIndexes, preNormalizeColumns,
} from './utils';

export class ColumnsController {
  private readonly columnsConfiguration: Subscribable<ColumnProperties[] | undefined>;
  private readonly columnsSettings: SubsGetsUpd<PreNormalizedColumn[]>;

  public readonly columns: SubsGets<Column[]>;

  public readonly visibleColumns: SubsGets<Column[]>;

  public readonly nonVisibleColumns: SubsGets<Column[]>;

  public readonly allowColumnReordering: Subscribable<boolean>;

  public static dependencies = [OptionsController, DataController] as const;

  constructor(
    private readonly options: OptionsController,
    private readonly dataController: DataController,
  ) {
    this.columnsConfiguration = this.options.oneWay('columns');

    const columnsFromDataSource = computed(
      (items: unknown[]) => {
        if (!items.length) {
          return [];
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return Object.keys(items[0] as any);
      },
      [this.dataController.items],
    );

    this.columnsSettings = interruptableComputed(
      (columnsConfiguration) => preNormalizeColumns(columnsConfiguration ?? []),
      [
        iif(
          computed((columnsConfiguration) => !!columnsConfiguration, [this.columnsConfiguration]),
          this.columnsConfiguration,
          columnsFromDataSource,
        ),
      ],
    );

    this.columns = computed(
      (columnsSettings) => normalizeColumns(columnsSettings ?? []),
      [
        this.columnsSettings,
      ],
    );

    this.visibleColumns = computed(
      (columns) => columns
        .filter((column): column is VisibleColumn => column.visible)
        .sort((a, b) => a.visibleIndex - b.visibleIndex),
      [this.columns],
    );

    this.nonVisibleColumns = computed(
      (columns) => columns.filter((column) => !column.visible),
      [this.columns],
    );

    this.allowColumnReordering = this.options.oneWay('allowColumnReordering');
  }

  public createDataRow(data: DataObject, columns: Column[]): DataRow {
    return {
      cells: columns.map((c) => {
        const displayValue = c.calculateDisplayValue(data);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        let text = formatHelper.format(displayValue as any, c.format);

        if (c.customizeText) {
          text = c.customizeText({
            value: displayValue,
            valueText: text,
          });
        }

        return {
          column: c,
          value: c.calculateCellValue(data),
          displayValue,
          text,
        };
      }),
      key: this.dataController.getDataKey(data),
      data,
    };
  }

  public addColumn(columnProps: ColumnProperties): void {
    this.columnsSettings.updateFunc((columns) => preNormalizeColumns([
      ...columns,
      columnProps,
    ]));
  }

  public deleteColumn(column: Column): void {
    this.columnsSettings.updateFunc(
      (columns) => columns.filter((c) => c.name !== column.name),
    );
  }

  public columnOption<TProp extends keyof ColumnSettings>(
    column: Column,
    option: TProp,
    value: ColumnSettings[TProp],
  ): void {
    this.columnsSettings.updateFunc((columns) => {
      const index = getColumnIndexByName(columns, column.name);
      const newColumns = [...columns];

      if (columns[index][option] === value) {
        return columns;
      }

      newColumns[index] = {
        ...newColumns[index],
        [option]: value,
      };

      const visibleIndexes = normalizeVisibleIndexes(
        newColumns.map((c) => c.visibleIndex),
        index,
      );

      visibleIndexes.forEach((visibleIndex, i) => {
        if (newColumns[i].visibleIndex !== visibleIndex) {
          newColumns[i] = {
            ...newColumns[i],
            visibleIndex,
          };
        }
      });

      return newColumns;
    });
  }
}
