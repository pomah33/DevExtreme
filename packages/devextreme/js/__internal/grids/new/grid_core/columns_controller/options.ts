/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { DataType } from '@js/common';
import messageLocalization from '@js/localization/message';

import type { DataObject } from '../data_controller/types';
import type { WithRequired } from '../types';
import type { Column } from './types';

export type ColumnSettings = Partial<Omit<Column, 'calculateDisplayValue'> & {
  calculateDisplayValue: string | ((this: Column, data: DataObject) => unknown);
}>;

export type PreNormalizedColumn = WithRequired<ColumnSettings, 'name' | 'visibleIndex'>;

export type ColumnProperties = ColumnSettings | string;

export const defaultColumnProperties = {
  dataType: 'string',
  calculateCellValue(data): unknown {
    return data[this.dataField!];
  },
  calculateDisplayValue(data): unknown {
    return this.calculateCellValue(data);
  },
  alignment: 'left',
  visible: true,
  allowReordering: true,
  trueText: messageLocalization.format('dxDataGrid-trueText'),
  falseText: messageLocalization.format('dxDataGrid-falseText'),
} satisfies Partial<Column>;

export const defaultColumnPropertiesByDataType: Record<
DataType,
Exclude<ColumnProperties, string>
> = {
  boolean: {
    alignment: 'center',
    customizeText({ value }): string {
      return value
        ? this.trueText
        : this.falseText;
    },
  },
  string: {

  },
  date: {

  },
  datetime: {

  },
  number: {
    alignment: 'right',
  },
  object: {

  },
};

export interface Options {
  columns?: ColumnProperties[];

  allowColumnReordering?: boolean;
}

export const defaultOptions = {
  allowColumnReordering: false,
} satisfies Options;
