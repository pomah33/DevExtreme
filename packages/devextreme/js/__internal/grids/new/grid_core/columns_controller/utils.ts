import { compileGetter } from '@js/core/utils/data';
import { captionize } from '@js/core/utils/inflector';
import { isDefined, isString } from '@js/core/utils/type';

import type { DataObject } from '../data_controller/types';
import type { ColumnProperties, ColumnSettings, PreNormalizedColumn } from './options';
import { defaultColumnProperties, defaultColumnPropertiesByDataType } from './options';
import type { Column } from './types';

function normalizeColumn(column: PreNormalizedColumn): Column {
  const dataTypeDefault = defaultColumnPropertiesByDataType[
    column.dataType ?? defaultColumnProperties.dataType
  ];

  const caption = captionize(column.name);

  const colWithDefaults = {
    ...defaultColumnProperties,
    ...dataTypeDefault,
    caption,
    ...column,
  };

  return {
    ...colWithDefaults,
    calculateDisplayValue: isString(colWithDefaults.calculateDisplayValue)
      ? compileGetter(colWithDefaults.calculateDisplayValue) as (data: DataObject) => string
      : colWithDefaults.calculateDisplayValue,
  };
}

export function getVisibleIndexes(
  indexes: (number | undefined)[],
): number[] {
  const newIndexes = [...indexes];
  let minNonExistingIndex = 0;

  indexes.forEach((visibleIndex, index) => {
    while (newIndexes.includes(minNonExistingIndex)) {
      minNonExistingIndex += 1;
    }

    newIndexes[index] = visibleIndex ?? minNonExistingIndex;
  });

  return newIndexes as number[];
}

export function normalizeVisibleIndexes(
  indexes: number[],
  forceIndex?: number,
): number[] {
  const indexMap = indexes.map(
    (visibleIndex, index) => [index, visibleIndex],
  );

  const sortedIndexMap = new Array<number>(indexes.length);
  if (isDefined(forceIndex)) {
    sortedIndexMap[indexes[forceIndex]] = forceIndex;
  }

  let j = 0;
  indexMap
    .sort((a, b) => a[1] - b[1])
    .forEach(([index]) => {
      if (index === forceIndex) {
        return;
      }

      if (isDefined(sortedIndexMap[j])) {
        j += 1;
      }

      sortedIndexMap[j] = index;
      j += 1;
    });

  const returnIndexes = new Array<number>(indexes.length);
  sortedIndexMap.forEach((index, visibleIndex) => {
    returnIndexes[index] = visibleIndex;
  });
  return returnIndexes;
}

export function normalizeColumns(columns: PreNormalizedColumn[]): Column[] {
  const normalizedColumns = columns.map((c) => normalizeColumn(c));
  return normalizedColumns;
}

export function preNormalizeColumns(columns: ColumnProperties[]): PreNormalizedColumn[] {
  const normalizedColumns = columns
    .map((column): ColumnSettings => {
      if (typeof column === 'string') {
        return {
          dataField: column,
        };
      }

      return column;
    })
    .map((column, index) => ({
      ...column,
      name: column.name ?? column.dataField ?? `column-${index}`,
    }));

  const visibleIndexes = getVisibleIndexes(
    normalizedColumns.map((c) => c.visibleIndex),
  );

  normalizedColumns.forEach((_, i) => {
    normalizedColumns[i].visibleIndex = visibleIndexes[i];
  });

  return normalizedColumns as PreNormalizedColumn[];
}

export function normalizeStringColumn(column: ColumnProperties): ColumnSettings {
  if (typeof column === 'string') {
    return { dataField: column };
  }

  return column;
}

export function getColumnIndexByName(columns: PreNormalizedColumn[], name: string): number {
  return columns.findIndex((c) => c.name === name);
}

export function getColumnByIndexOrName(
  columns: Column[],
  columnNameOrIndex: string | number,
): Column | undefined {
  const column = columns.find((c, i) => {
    if (isString(columnNameOrIndex)) {
      return c.name === columnNameOrIndex;
    }
    return i === columnNameOrIndex;
  });

  return column;
}
