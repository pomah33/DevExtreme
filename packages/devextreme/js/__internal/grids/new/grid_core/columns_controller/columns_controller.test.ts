/* eslint-disable spellcheck/spell-checker */
import { describe, expect, it } from '@jest/globals';

import { DataController } from '../data_controller';
import type { Options } from '../options';
import { OptionsControllerMock } from '../options_controller/options_controller.mock';
import { ColumnsController } from './columns_controller';

const setup = (config: Options = {}) => {
  const options = new OptionsControllerMock(config);

  const dataController = new DataController(options);

  const columnsController = new ColumnsController(options, dataController);

  return {
    options,
    dataController,
    columnsController,
  };
};

describe('ColumnsController', () => {
  describe('columns', () => {
    it('should contain processed column configs', () => {
      const { columnsController } = setup({
        columns: [
          'a',
          { dataField: 'b' },
          { dataField: 'c', visible: false },
        ],
      });

      const columns = columnsController.columns.unreactive_get();
      expect(columns).toMatchSnapshot();
    });
  });
  describe('visibleColumns', () => {
    it('should contain visible columns', () => {
      const { columnsController } = setup({
        columns: [
          'a',
          { dataField: 'b' },
          { dataField: 'c', visible: false },
        ],
      });

      const visibleColumns = columnsController.visibleColumns.unreactive_get();
      expect(visibleColumns).toHaveLength(2);
      expect(visibleColumns[0].name).toBe('a');
      expect(visibleColumns[1].name).toBe('b');
    });
  });
  describe('nonVisibleColumns', () => {
    it('should contain non visible columns', () => {
      const { columnsController } = setup({
        columns: [
          'a',
          { dataField: 'b' },
          { dataField: 'c', visible: false },
        ],
      });

      const nonVisibleColumns = columnsController.nonVisibleColumns.unreactive_get();
      expect(nonVisibleColumns).toHaveLength(1);
      expect(nonVisibleColumns[0].name).toBe('c');
    });
  });

  describe('createDataRow', () => {
    it('should process data object to data row using column configuration', () => {
      const { columnsController } = setup({
        columns: [
          'a',
          { dataField: 'b' },
        ],
      });

      const columns = columnsController.columns.unreactive_get();
      const dataObject = { a: 'my a value', b: 'my b value' };
      const dataRow = columnsController.createDataRow(dataObject, columns);
      expect(dataRow).toMatchSnapshot();
    });
  });

  describe('addColumn', () => {
    it('should add new column to columns', () => {
      const { columnsController } = setup(
        { columns: ['a', 'b'] },
      );

      let columns = columnsController.columns.unreactive_get();
      expect(columns).toHaveLength(2);
      expect(columns).toMatchObject([
        { dataField: 'a' },
        { dataField: 'b' },
      ]);

      columnsController.addColumn('c');

      columns = columnsController.columns.unreactive_get();
      expect(columns).toHaveLength(3);
      expect(columns).toMatchObject([
        { dataField: 'a' },
        { dataField: 'b' },
        { dataField: 'c' },
      ]);
    });
  });

  describe('deleteColumn', () => {
    it('should remove given column from columns', () => {
      const { columnsController } = setup(
        { columns: ['a', 'b'] },
      );

      let columns = columnsController.columns.unreactive_get();
      expect(columns).toHaveLength(2);
      expect(columns).toMatchObject([
        { dataField: 'a' },
        { dataField: 'b' },
      ]);

      columnsController.deleteColumn(columns[1]);

      columns = columnsController.columns.unreactive_get();
      expect(columns).toHaveLength(1);
      expect(columns).toMatchObject([
        { dataField: 'a' },
      ]);
    });
  });

  describe('columnOption', () => {
    it('should update option of given column', () => {
      const { columnsController } = setup(
        { columns: ['a', 'b'] },
      );

      let columns = columnsController.columns.unreactive_get();
      expect(columns).toMatchObject([
        { dataField: 'a', visible: true },
        { dataField: 'b', visible: true },
      ]);

      columnsController.columnOption(columns[1], 'visible', false);

      columns = columnsController.columns.unreactive_get();
      expect(columns).toMatchObject([
        { dataField: 'a', visible: true },
        { dataField: 'b', visible: false },
      ]);
    });

    it('should correctly update visibleIndex option for all columns', () => {
      const { columnsController } = setup(
        { columns: ['a', 'b', 'c'] },
      );

      let columns = columnsController.columns.unreactive_get();
      expect(columns).toMatchObject([
        { dataField: 'a', visibleIndex: 0 },
        { dataField: 'b', visibleIndex: 1 },
        { dataField: 'c', visibleIndex: 2 },
      ]);

      columnsController.columnOption(columns[2], 'visibleIndex', 0);

      columns = columnsController.columns.unreactive_get();
      expect(columns).toMatchObject([
        { dataField: 'a', visibleIndex: 1 },
        { dataField: 'b', visibleIndex: 2 },
        { dataField: 'c', visibleIndex: 0 },
      ]);
    });
  });
});
