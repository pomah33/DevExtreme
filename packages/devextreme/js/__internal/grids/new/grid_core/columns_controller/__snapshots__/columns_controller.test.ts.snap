// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnsController columns should contain processed column configs 1`] = `
[
  {
    "alignment": "left",
    "allowReordering": true,
    "calculateCellValue": [Function],
    "calculateDisplayValue": [Function],
    "caption": "A",
    "dataField": "a",
    "dataType": "string",
    "falseText": "false",
    "name": "a",
    "trueText": "true",
    "visible": true,
    "visibleIndex": 0,
  },
  {
    "alignment": "left",
    "allowReordering": true,
    "calculateCellValue": [Function],
    "calculateDisplayValue": [Function],
    "caption": "B",
    "dataField": "b",
    "dataType": "string",
    "falseText": "false",
    "name": "b",
    "trueText": "true",
    "visible": true,
    "visibleIndex": 1,
  },
  {
    "alignment": "left",
    "allowReordering": true,
    "calculateCellValue": [Function],
    "calculateDisplayValue": [Function],
    "caption": "C",
    "dataField": "c",
    "dataType": "string",
    "falseText": "false",
    "name": "c",
    "trueText": "true",
    "visible": false,
    "visibleIndex": 2,
  },
]
`;

exports[`ColumnsController createDataRow should process data object to data row using column configuration 1`] = `
{
  "cells": [
    {
      "column": {
        "alignment": "left",
        "allowReordering": true,
        "calculateCellValue": [Function],
        "calculateDisplayValue": [Function],
        "caption": "A",
        "dataField": "a",
        "dataType": "string",
        "falseText": "false",
        "name": "a",
        "trueText": "true",
        "visible": true,
        "visibleIndex": 0,
      },
      "displayValue": "my a value",
      "text": "my a value",
      "value": "my a value",
    },
    {
      "column": {
        "alignment": "left",
        "allowReordering": true,
        "calculateCellValue": [Function],
        "calculateDisplayValue": [Function],
        "caption": "B",
        "dataField": "b",
        "dataType": "string",
        "falseText": "false",
        "name": "b",
        "trueText": "true",
        "visible": true,
        "visibleIndex": 1,
      },
      "displayValue": "my b value",
      "text": "my b value",
      "value": "my b value",
    },
  ],
  "data": {
    "a": "my a value",
    "b": "my b value",
  },
  "key": {
    "a": "my a value",
    "b": "my b value",
  },
}
`;
